In [1]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [2]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [3]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [4]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [33]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
market_data_scaled = StandardScaler().fit_transform(df_market_data)	

In [34]:
# Create a DataFrame with the scaled data
df_market_data_scaled = pd.DataFrame(market_data_scaled, columns=['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y'])

# Copy the crypto names from the original data
df_market_data_scaled['coin_id'] = df_market_data.index

# Set the coinid column as index
df_market_data_scaled = df_market_data_scaled.set_index('coin_id')

# Display sample data
df_market_data_scaled.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [35]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1, 11))

In [40]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    k_model = KMeans(n_clusters=i, random_state=0)
    k_model.fit(df_market_data_scaled)
    inertia.append(k_model.inertia_)
    


c:\Users\trans\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\cluster\_kmeans.py:1416: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
c:\Users\trans\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\cluster\_kmeans.py:1416: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
c:\Users\trans\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\cluster\_kmeans.py:1416: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
c:\Users\trans\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\clus

In [41]:
# Create a dictionary with the data to plot the Elbow curve
elbow = {"k": k, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow)
df_elbow

k     inertia
0   1  287.000000
1   2  198.571818
2   3  123.190482
3   4   79.022435
4   5   65.302379
5   6   52.101806
6   7   46.035250
7   8   37.284807
8   9   32.161417
9  10   28.075253

In [45]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
k_means_elbow = df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="Inertia by K", 
    xticks=k, 
    color = 'green', 
)
k_means_elbow

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [46]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=0)

In [47]:
# Fit the K-Means model using the scaled data
model.fit(df_market_data_scaled)

c:\Users\trans\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\cluster\_kmeans.py:1416: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)


KMeans(n_clusters=4, random_state=0)

In [48]:
# Predict the clusters to group the cryptocurrencies using the scaled data
kmeans_prediction = model.predict(df_market_data_scaled)

# Print the resulting array of cluster values.
kmeans_prediction

array([2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 1, 2, 0, 0, 3, 0, 0, 0, 0])

In [52]:
# Create a copy of the DataFrame
df_market_predictions = df_market_data.copy()

In [53]:
# Add a new column to the DataFrame with the predicted clusters
df_market_predictions['cluster'] = kmeans_prediction

# Display sample data
df_market_predictions

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                   1.08388                     7.60278   
ethereum                                  0.22392                    10.38134   
tether                                   -0.21173                     0.04935   
ripple                                   -0.37819                    -0.60926   
bitcoin-cash                              2.90585                    17.09717   
binancecoin                               2.10423                    12.85511   
chainlink                                -0.23935                    20.69459   
cardano                                   0.00322                    13.99302   
litecoin                                 -0.06341                     6.60221   
bitcoin-cash-sv                           0.92530                     3.29641   
crypto-com-chain                          0.61209                    -5.67151   
usd-coin                                 -0.17825                    -0.11871   
eos                                       0.14477                    -1.31177   
monero                                    0.42961                    15.78515   
tron                                      0.07647                     4.23886   
tezos                                    -0.67316                     8.95665   
okb                                      -2.72700                    -4.55389   
stellar                                  -1.00843                     2.07149   
cosmos                                   -0.95103                    16.08534   
cdai                                      0.21169                     0.05820   
neo                                       0.49302                     2.44243   
wrapped-bitcoin                           1.10231                     7.40537   
leo-token                                -0.13192                    -1.34886   
huobi-token                              -0.40818                     1.61798   
nem                                      -0.84990                    -0.85140   
binance-usd                              -0.10642                     0.04726   
iota                                      0.41996                     6.06830   
vechain                                   1.28766                    -1.76352   
zcash                                    -0.60897                    10.34780   
theta-token                              -4.56089                    -6.09456   
dash                                     -1.06006                     5.09387   
ethereum-classic                         -0.45950                     3.05209   
ethlend                                 -13.52786                     4.21266   
maker                                    -0.60285                     8.15400   
havven                                   -4.07216                     4.33651   
omisego                                   4.84033                     6.82985   
celsius-degree-token                      2.51323                     0.60354   
ontology                                 -1.35845                    -1.21399   
ftx-token                                 0.83416                     7.10438   
true-usd                                 -0.06197                     0.16642   
digibyte                                  2.97084                     0.67038   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                   6.57509   
ethereum                                  4.80849   
tether                                    0.00640   
ripple                                    2.24984   
bitcoin-cash                             14.75334   
binancecoin                               6.80688   
chainlink                                 9.30098   
cardano                                   5.55476   
litecoin

In [54]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
k_means_scatter = df_market_predictions.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="cluster",
    hover_cols = ["coin_id"], 
    title = "Scaled Data Clusters"
)
k_means_scatter

:NdOverlay   [cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [57]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

PCA(n_components=3)

In [61]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
market_data_pca = pca.fit_transform(df_market_data_scaled)
# View the first five rows of the DataFrame. 
market_data_pca[0:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [62]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
explained_variance_ratio = pca.explained_variance_ratio_
print(explained_variance_ratio)

[0.3719856  0.34700813 0.17603793]


In [63]:
#add the explained variance 
sum(explained_variance_ratio)

0.895031657030984

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:**  89.5%

In [66]:
# Create a new DataFrame with the PCA data.
df_market_pca = pd.DataFrame(market_data_pca, columns = ["PC1","PC2","PC3"])

# Copy the crypto names from the original data
df_market_pca['coin_id'] = df_market_data.index

# Set the coinid column as index
df_market_pca = df_market_pca.set_index('coin_id')

# Display sample data
df_market_pca.head()

PC1       PC2       PC3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [21]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1, 11))

In [67]:
# Create an empty list to store the inertia values
inertia_pca = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i in k: 
    k_model = KMeans(n_clusters=i, random_state=0)
    k_model.fit(df_market_pca)
    inertia_pca.append(k_model.inertia_)

c:\Users\trans\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\cluster\_kmeans.py:1416: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
c:\Users\trans\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\cluster\_kmeans.py:1416: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
c:\Users\trans\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\cluster\_kmeans.py:1416: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
c:\Users\trans\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\clus

In [68]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_pca = {"k": k, "inertia": inertia_pca}
# Create a DataFrame with the data to plot the Elbow curve
df_elbow_pca = pd.DataFrame(elbow_data_pca)
df_elbow_pca

k     inertia
0   1  256.874086
1   2  168.811896
2   3   93.774626
3   4   49.665497
4   5   37.839466
5   6   27.720626
6   7   21.182776
7   8   17.016962
8   9   13.668325
9  10   10.484890

In [70]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
k_means_elbow_pca = df_elbow_pca.hvplot.line(
    x="k", 
    y="inertia", 
    title="PCA Elbow Curve", 
    xticks=k, 
    color = 'red'
)
k_means_elbow_pca

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No, both are 4.

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [71]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=0)

In [72]:
# Fit the K-Means model using the PCA data
model.fit(df_market_pca)

c:\Users\trans\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\cluster\_kmeans.py:1416: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)


KMeans(n_clusters=4, random_state=0)

In [73]:
# Predict the clusters to group the cryptocurrencies using the PCA data
kmeans_pca_predictions = model.predict(df_market_pca)
# Print the resulting array of cluster values.
kmeans_pca_predictions

array([2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 1, 2, 0, 0, 3, 0, 0, 0, 0])

In [75]:
# Create a copy of the DataFrame with the PCA data
df_market_pca_predictions = df_market_pca.copy()

# Add a new column to the DataFrame with the predicted clusters
df_market_pca_predictions['cluster_predictions']=kmeans_pca_predictions

# Display sample data
df_market_pca_predictions

PC1       PC2       PC3  cluster_predictions
coin_id                                                                
bitcoin              -0.600667  0.842760  0.461595                    2
ethereum             -0.458261  0.458466  0.952877                    2
tether               -0.433070 -0.168126 -0.641752                    0
ripple               -0.471835 -0.222660 -0.479053                    0
bitcoin-cash         -1.157800  2.041209  1.859715                    2
binancecoin          -0.516534  1.388377  0.804071                    2
chainlink            -0.450711  0.517699  2.846143                    2
cardano              -0.345600  0.729439  1.478013                    2
litecoin             -0.649468  0.432165  0.600303                    2
bitcoin-cash-sv      -0.759014 -0.201200 -0.217653                    0
crypto-com-chain     -0.248198 -1.376252 -1.462026                    0
usd-coin             -0.438408 -0.175337 -0.663388                    0
eos                  -0.693425 -0.473815 -0.527597                    0
monero                0.060499  2.909404  1.498571                    2
tron                 -0.393352 -0.108192 -0.012756                    0
tezos                -0.796176 -0.494409  1.082812                    2
okb                   0.064075 -1.269825 -1.098829                    0
stellar              -0.489015 -0.732719 -0.062543                    0
cosmos               -0.306272  0.703415  1.714224                    2
cdai                 -0.513528 -0.142802 -0.656566                    0
neo                  -0.362120 -0.986914 -0.728752                    0
wrapped-bitcoin      -0.604265  0.827398  0.439316                    2
leo-token            -0.413296 -0.674115 -1.076628                    0
huobi-token          -0.407483 -0.212507 -0.351426                    0
nem                   0.608974  0.563532 -1.148742                    0
binance-usd          -0.450211 -0.151019 -0.647401                    0
iota                 -0.764665 -0.517886  0.204990                    0
vechain              -0.556315 -1.938209 -1.261776                    0
zcash                -0.425147  0.492976  1.058048                    2
theta-token           2.676868 -0.013954 -1.965207                    0
dash                 -0.613923 -0.479337  0.339565                    0
ethereum-classic     -0.579924 -0.356334 -0.114942                    0
ethlend               8.089018 -3.896891  2.301382                    1
maker                -0.389045  0.165041  0.379414                    2
havven                0.865762 -2.261882  0.275583                    0
omisego               0.111675  0.428316 -1.205398                    0
celsius-degree-token  4.792395  6.767679 -1.986985                    3
ontology             -0.632355 -2.108117 -0.652227                    0
ftx-token            -0.593142  0.021485  0.209911                    0
true-usd             -0.458131 -0.135734 -0.635284                    0
digibyte             -0.297910 -0.191126 -0.909602                    0

In [78]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
k_means_scatter_pca = df_market_pca_predictions.hvplot.scatter(
    x="PC1",
    y="PC2",
    by="cluster_predictions",
    hover_cols = ["coin_id"], 
    title = "PCA Clusters"
)
k_means_scatter_pca

:NdOverlay   [cluster_predictions]
   :Scatter   [PC1]   (PC2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [79]:
# Composite plot to contrast the Elbow curves
(k_means_elbow + k_means_elbow_pca).cols(1)

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [80]:
# Composite plot to contrast the clusters
(k_means_scatter + k_means_scatter_pca).cols(1)

:Layout
   .NdOverlay.I  :NdOverlay   [cluster]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [cluster_predictions]
      :Scatter   [PC1]   (PC2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** Using fewer features with K-means clustering through PCA allows for clearer visualization of cluster distinctions. However, the optimal number of clusters continued to remain the same for both models at 4.